[distributed] Fix _ReaderView.read() and readinto() to stop reading at the end of the slice

mhorowitz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143360
* #143359
* #143358
* __->__ #143357

_ReaderView doesn't work correctly if the slice ends past the view.

read(-1) would call read(-1) on the base_stream, which would consume the entire underlying stream, even if the view ended before that.
read(n) would read n bytes, even if the view ended before that.

The new implementation clamps the size read to the size of the view.

readinto(b) would read len(b) bytes, even if the view ended before that.

Since the interface depends on the size of b, we use a (potentially) shortened view into b to avoid a copy.  If the view doesn't contain enough data to fill the view, then this will appear as end of stream to the caller, which is the desired behavior.

This fix should not be user facing, since the bug is in an internal helper, and is only visible with new code down the stack.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @pradeepfn @ekr0

mhorowitz committed:
[distributed] Fix _ReaderView.read() and readinto() to stop reading at the end of the slice

[ghstack-poisoned]

mhorowitz committed:
Update on "[distributed] Fix _ReaderView.read() and readinto() to stop reading at the end of the slice"




[ghstack-poisoned]

mhorowitz committed:
Update on "[distributed] Fix _ReaderView.read() and readinto() to stop reading at the end of the slice"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn

[ghstack-poisoned]

saumishr commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -365,9 +365,19 @@ def seekable(self) -> bool:
 
     def read(self, size=-1):
+        max_size = self.len - self.tell()
+        if size == -1:
```
`minor nit`
`
if size == -1 or size > max_size:
`

saumishr reviewed (commented):
Looks good to me

mhorowitz committed:
Update on "[distributed] Fix _ReaderView.read() and readinto() to stop reading at the end of the slice"


_ReaderView doesn't work correctly if the slice ends past the view.

read(-1) would call read(-1) on the base_stream, which would consume the entire underlying stream, even if the view ended before that.
read(n) would read n bytes, even if the view ended before that.

The new implementation clamps the size read to the size of the view.

readinto(b) would read len(b) bytes, even if the view ended before that.

Since the interface depends on the size of b, we use a (potentially) shortened view into b to avoid a copy.  If the view doesn't contain enough data to fill the view, then this will appear as end of stream to the caller, which is the desired behavior.

This fix should not be user facing, since the bug is in an internal helper, and is only visible with new code down the stack.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn

[ghstack-poisoned]

mhorowitz commented:
@pytorchbot merge

saumishr reviewed (approved):
Changes look good to me. Thanks for the UTs. Approving the PR assuming that we have validated this change end to end with a prod model as well.

mhorowitz commented:
@pytorchbot merge
